### PR TITLE
Implementation of an interactive game

### PR DESCRIPTION
**Pull request by d9nchik**
# Interactive game
- Implement basic scenarios
- Integrate remote data source
- Replace static data with real work with the database

### CODE REVIEW

File: frontend/src/components/game/GameBody.tsx [hunk 1]
```diff
@@ -15,31 +18,29 @@ const GameBody: FunctionComponent<IProps> = ({ gameStatus }: IProps) => {
...
   >('');
 
-  const daysInUniversityString = `${daysInUniversity} ${
-    daysInUniversity === 1
-      ? 'день'
-      : daysInUniversity <= 4 && daysInUniversity > 1
-      ? 'дня'
-      : 'дней'
-  }`;
-
```
Comment (d9nchik) [inline frontend/src/components/game/GameBody.tsx]: Cool, my code has been deleted:angry:

File: backend/functions/src/index.ts [hunk 2]
```diff
@@ -1,9 +1,73 @@
...
+    res.header('Access-Control-Allow-Origin', '*'); // update to match the domain you will make the request from
+    res.header(
+      'Access-Control-Allow-Headers',
+      'Origin, X-Requested-With, Content-Type, Accept'
+    );
+
+    try {
+      functions.logger.log(`rawBody = ${String(req.rawBody)}`);
+
+      const uid = req.body.data.uid as string;
```
Comment (nechaido) [inline backend/functions/src/index.ts]:
```suggestion
      if (!req.body || !req.body.data || req.body.data.uid !== 'string')
        handleError();;
      const { uid } = req.body.data;
```

Comment (d9nchik) [reply backend/functions/src/index.ts]: Thanks, I have added this refactoring with commits:
- [reduce amount of useless try/catch in backend](https://github.com/d9nchik/kpi_student/pull/6/commits/90cda29c69a430f5b4141dba6b67cd690a24dfd1)
- [remove redundant 'as' at backend/getRandomQuiz](https://github.com/d9nchik/kpi_student/pull/6/commits/2699399cba31d3da7c18b6cc53d77b1e0b4f8dd4)

File: frontend/src/data/auth.ts [hunk 3]
```diff
@@ -83,3 +58,14 @@ auth.onAuthStateChanged(newUser => {
 export const isAuthenticated = (): boolean => !!user;
 
 export const getUser = (): firebase.User | null => user;
+
+export async function executeAsyncFunctionIfErrReturnMessage(
+  inputFunction: () => Promise<firebase.auth.UserCredential | void>
+): Promise<string | null> {
```
Comment (nechaido) [inline frontend/src/data/auth.ts]: Type alias
```suggestion
): Promise<errMsg | null> {
```

Comment (nechaido) [reply frontend/src/data/auth.ts]: Or Error instance.

File: frontend/src/data/dataStorage.ts [hunk 4]
```diff
@@ -110,6 +111,9 @@ export const addQuiz = async (quiz: QuizWithOnlyBody): Promise<boolean> => {
     },
     imageURL: '',
   };
+  quizWithBody.answerVariants = quizWithBody.answerVariants.map(
```
Comment (nechaido) [inline frontend/src/data/dataStorage.ts]:
```suggestion
  const quizeNormalized = { ...quizWithBody, answerVariants: quizWithBody.answerVariants.map( 
```

File: frontend/src/data/auth.ts (hunk 3 above)
Comment (d9nchik) [reply frontend/src/data/auth.ts]: Thanks, fixed with commits:
- [change executeAsyncFunctionIfErrReturnMessage in auth.ts return type](https://github.com/d9nchik/kpi_student/pull/6/commits/0c3025ad53b7d5423dee5214bb0afa5b5f8c7ee4)

File: frontend/src/data/dataStorage.ts (hunk 4 above)
Comment (d9nchik) [reply frontend/src/data/dataStorage.ts]: Thanks, fixed with commit: [fix addQuiz to recreate new object and not to modify old](https://github.com/d9nchik/kpi_student/pull/6/commits/b3e750743a203c714a32f23f38c4fd7d88544e81)

**Pull request merged**